methods/guest/src/bin/email_with_regex_verify.rs

```diff
@@ -0,0 +1,102 @@
+#![no_main]
+
+use cfdkim::{ verify_email_with_key, DkimPublicKey };
+use mailparse::{ parse_mail, ParsedMail };
+use risc0_zkvm::guest::env;
+use sha2::{ Digest, Sha256 };
+use slog::{ o, Discard, Logger };
+use zkemail_core::{ EmailVerifierOutput, EmailWithRegex, EmailWithRegexVerifierOutput, DFA };
+use regex_automata::{ dfa::{ dense, regex::Regex }, Match };
+
+fn hash_bytes(data: &[u8]) -> Vec<u8> {
+    let mut hasher = Sha256::new();
+    hasher.update(data);
+    hasher.finalize().to_vec()
+}
+
+fn extract_email_body(parsed_email: &ParsedMail) -> String {
+    if parsed_email.subparts.is_empty() {
+        return parsed_email.get_body().unwrap();
+    }
+
+    let body = parsed_email.subparts
+        .iter()
+        .find(|part| part.ctype.mimetype == "text/html")
+        .map(|part| part.get_body())
+        .unwrap_or_else(|| parsed_email.subparts[0].get_body());
+
+    body.unwrap()
+}
+
+fn process_regex_parts(parts: &[(bool, DFA)], email_body: &str) -> (bool, Vec<String>) {
+    let mut regex_verified = false;
+    let mut regex_matches = Vec::with_capacity(
+        parts
+            .iter()
+            .filter(|(is_public, _)| *is_public)
+            .count()
+    );
+
+    for part in parts {
+        let (is_public, dfa) = part;
+        let fwd = dense::DFA::from_bytes(&dfa.fwd).unwrap().0;
+        let rev = dense::DFA::from_bytes(&dfa.bwd).unwrap().0;
+
+        let re = Regex::builder().build_from_dfas(fwd, rev);
+        let matches: Vec<Match> = re.find_iter(email_body).collect();
+
+        if !matches.is_empty() {
+            regex_verified = true;
+            if *is_public {
+                let substring = email_body[matches[0].start()..matches[0].end()].to_string();
+                regex_matches.push(substring);
+            }
+        }
```

Comment:
@shreyas-londhe 
Could you explain why these processes are enough to verify the regex?
Specifically, what are roles of `fwd` and `rev` here?

Reply:
There is a bug in the regex matching which I'm fixing, and will reflect in #2.

> Specifically, what are roles of `fwd` and `rev` here?
`fwd` and `rev` are the DFAs serialized out of the circuit and are fed to the circuit as bytes. In the circuit we just deserialize it and construct the Regex which is the most efficient for matching ops. 

Reference: https://docs.rs/regex-automata/latest/regex_automata/dfa/regex/struct.Builder.html#method.build_from_dfas